Lossy Numeric Regex

ivakegg opened the pull request:
original regex for number types

ivakegg committed:
Removed the renormalization of terms when pulling out of the query for
expansion

jack-gitdev commented on warehouse/query-core/src/test/java/datawave/query/util/WiseGuysIngest.java:
```diff
@@ -127,6 +129,7 @@ public static void writeItAll(AccumuloClient client, WhatKindaRange range) throw
             mutation.put(datatype + "\u0000" + corleoneChildUID, "BIRTH_DATE" + "\u0000" + "1930-12-28T00:00:05.000Z", columnVisibility, timeStamp, emptyValue);
             mutation.put(datatype + "\u0000" + corleoneChildUID, "DEATH_DATE" + "\u0000" + "2000-12-28T00:00:05.000Z", columnVisibility, timeStamp, emptyValue);
 
+            mutation.put(datatype + "\u0000" + sopranoUID, "BAIL.0" + "\u0000" + "123401", columnVisibility, timeStamp, emptyValue);
```
Nice catch on this edge case!

ivakegg committed:
Updated the normalization to include an EvalOnly clause with the original regex for number types
Updated type-utils submodule

ivakegg committed:
Merge branch 'integration' into bugfix/avoidMultiNormalize

ivakegg committed:
Simplified logic, and removed unused parameter

ivakegg committed:
Merge branch 'bugfix/avoidMultiNormalize' of github.com:NationalSecurityAgency/datawave into bugfix/avoidMultiNormalize

ivakegg committed:
Merge branch 'integration' into feature/numericRegex

ivakegg committed:
Merge branch 'integration' into feature/numericRegex

ivakegg committed:
redundant method

ivakegg committed:
fix type

hlgp commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java:
```diff
@@ -317,10 +318,15 @@ protected JexlNode expandNodeForNormalizers(JexlNode node, Object data) {
                             if (normalizer instanceof OneToManyNormalizerType && ((OneToManyNormalizerType<?>) normalizer).expandAtQueryTime()) {
+                                if (regexNode) {
+                                    throw new IllegalArgumentException(
+                                                    "OneToManyNormalizers to not handle regex normalization: " + fieldName + " -> " + normalizer.getClass());
```
this is no longer accurate.  there are some one-to-many normalizers which can and probably should now.

hlgp committed:
corresponding updates for type-utils zero regex checks
if a regex is lossy add the eval_only node to the group of nodes created by looping through all normalizers

hlgp committed:
more adjustments

hlgp committed:
move lossy check outside contains check

hlgp committed:
unused imports

ivakegg committed:
Merge branch 'integration' into feature/numericRegex

ivakegg committed:
Updated to ensure regexes normalization is applied even to oneToMany
normalizers, fixed a bug wih a missing reference expression, and changed
eval only to be left alone.

ivakegg committed:
Merge branch 'integration' into feature/numericRegex

ivakegg committed:
Merge remote-tracking branch 'origin/feature/numericRegex_hp' into feature/numericRegex

ivakegg committed:
Merge branch 'bugfix/avoidMultiNormalize' into feature/numericRegex

ivakegg committed:
Updated to avoid normalization again when doing index lookups
(invalidates the eval only work and is redundant)

ivakegg committed:
Updated message

apmoriarty commented on warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java:
```diff
@@ -372,6 +383,77 @@ public void testMethodNoExpansion() throws Exception {
+        Map<String,String> extraParameters = new HashMap<>();
+        extraParameters.put("include.grouping.context", "true");
+        extraParameters.put("hit.list", "true");
+        // extraParameters.put("query.syntax", "LUCENE");
```
this can go away

ivakegg replied:
done

apmoriarty commented on warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java:
```diff
@@ -372,6 +383,77 @@ public void testMethodNoExpansion() throws Exception {
+        Map<String,String> extraParameters = new HashMap<>();
+        extraParameters.put("include.grouping.context", "true");
+        extraParameters.put("hit.list", "true");
+        // extraParameters.put("query.syntax", "LUCENE");
```
this can go away

ivakegg replied:
done

apmoriarty commented on warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java:
```diff
@@ -372,6 +383,77 @@ public void testMethodNoExpansion() throws Exception {
+        Map<String,String> extraParameters = new HashMap<>();
+        extraParameters.put("include.grouping.context", "true");
+        extraParameters.put("hit.list", "true");
+        // extraParameters.put("query.syntax", "LUCENE");
```
this can go away

ivakegg replied:
done

ivakegg committed:
Merge branch 'integration' into feature/numericRegex

hlgp commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java:
```diff
@@ -329,29 +331,55 @@ protected JexlNode expandNodeForNormalizers(JexlNode node, Object data) {
-                                        normalizedOneToManyNodes.add(JexlNodeFactory.buildUntypedNode(node, fieldName, iter.next()));
+                                    JexlNode normalizedNode = JexlNodeFactory.buildUntypedNode(node, fieldName, normTerm);
+                                    if ((!term.equals(normTerm)) && normalizer.normalizedRegexIsLossy(term)) {
+                                        JexlNode evalOnly = QueryPropertyMarker.create(JexlNodeFactory.buildUntypedNode(node, fieldName, term),
```
we need to check if the term is indexOnly before adding EVAL_ONLY node

ivakegg replied:
done

ivakegg committed:
Updated to avoid eval only with index-only fields

apmoriarty commented on warehouse/query-core/src/test/java/datawave/query/util/WiseGuysIngest.java:
```diff
@@ -429,6 +453,10 @@ public static void writeItAll(AccumuloClient client, WhatKindaRange range) throw
             mutation.put("SENTENCE", shard + "\u0000" + datatype, columnVisibility, timeStamp,
                             range == WhatKindaRange.SHARD ? getValueForNuthinAndYourHitsForFree() : getValueForBuilderFor(caponeUID));
             bw.addMutation(mutation);
+            mutation = new Mutation("+dE4.015");
```
I apologize for not catching this earlier. This should be `normalizer.normalize()` the numeric value, perhaps with the normalized value in a comment

apmoriarty reviewed (approved):
Look good, one non-blocking comment

hlgp reviewed (approved):
testing looked great
